core/templates/pages/teach-page/teach-page.component.spec.ts

```diff
@@ -15,112 +15,193 @@
+import { SiteAnalyticsService } from 'services/site-analytics.service';
+import { TranslateService } from 'services/translate.service';
+import { WindowRef } from 'services/contextual/window-ref.service';
 
-    var $rootScope = $injector.get('$rootScope');
-    $scope = $rootScope.$new();
+@Pipe({name: 'translate'})
+class MockTranslatePipe {
+  transform(value: string, params: Object | undefined):string {
+    return value;
+  }
+}
+
+class MockTranslateService {
+  languageCode = 'es';
+  use(newLanguageCode: string): string {
+    this.languageCode = newLanguageCode;
+    return this.languageCode;
+  }
+}
+
+class MockI18nLanguageCodeService {
+  codeChangeEventEmiiter = new EventEmitter<string>();
+  getCurrentI18nLanguageCode() {
+    return 'en';
+  }
+
+  get onI18nLanguageCodeChange() {
+    return this.codeChangeEventEmiiter;
+  }
+}
+
+// Mocking window object here because changing location.href causes the
+// full page to reload. Page reloads raise an error in karma.
+class MockWindowRef {
+  _window = {
+    location: {
+      _hash: '',
+      _hashChange: null,
+      _href: '',
+      get hash() {
+        return this._hash;
+      },
+      set hash(val) {
+        this._hash = val;
+        if (this._hashChange === null) {
+          return;
+        }
+        this._hashChange();
+      },
+      get href() {
+        return this._href;
+      },
+      set href(val) {
+        this._href = val;
+      },
+      reload: (val) => val
+    },
+    get onhashchange() {
+      return this.location._hashChange;
+    },
+
+    set onhashchange(val) {
+      this.location._hashChange = val;
+    }
+  };
+  get nativeWindow() {
+    return this._window;
+  }
+}
+
+class MockSiteAnalyticsService {
+  registerApplyToTeachWithOppiaEvent(): void {
+    return;
+  }
+}
+
+let component: TeachPageComponent;
+let fixture: ComponentFixture<TeachPageComponent>;
 
-    ctrl = $componentController('teachPage', {
-      $rootScope: $scope
-    });
-  }));
+describe('Teach Page', function() {
+  let windowRef: MockWindowRef;
+  let sas = null;
+  let i18n = null;
+  let translate = null;
+
+  beforeEach(() => {
+    windowRef = new MockWindowRef();
+    TestBed.configureTestingModule({
+      declarations: [TeachPageComponent, MockTranslatePipe],
+      providers: [
+        {
+          provide: I18nLanguageCodeService,
+          useClass: MockI18nLanguageCodeService
+        },
+        { provide: SiteAnalyticsService, useClass: MockSiteAnalyticsService },
+        { provide: TranslateService, useClass: MockTranslateService },
+        UrlInterpolationService,
+        { provide: WindowRef, useValue: windowRef }
+      ],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+    i18n = TestBed.get(I18nLanguageCodeService);
+    translate = TestBed.get(TranslateService);
+    sas = TestBed.get(SiteAnalyticsService);
+    fixture = TestBed.createComponent(TeachPageComponent);
```

Comment:
These test needs to be formatted a bit.

Declaration of `fixture` and `component` should be done in another `beforeEach` similar to  https://angular.io/guide/testing-components-basics#cli-generated-tests. You could refer `thanks-page.component.spec.ts` where I did the same way.

Reply:
Done